aptos-move/framework/aptos-framework/sources/aptos_governance.move

```diff
@@ -210,6 +237,49 @@ module aptos_framework::aptos_governance {
         borrow_global<GovernanceConfig>(@aptos_framework).required_proposer_stake
     }
 
+    #[view]
+    /// Return true if a stake pool has already voted on a proposal before partial governance voting is enabled.
+    public fun has_entirely_voted(stake_pool: address, proposal_id: u64): bool acquires VotingRecords {
+        let record_key = RecordKey {
+            stake_pool,
+            proposal_id,
+        };
+        // If a stake pool has already voted on a proposal before partial governance voting is enabled,
+        // there is a record in VotingRecords.
+        let voting_records = borrow_global<VotingRecords>(@aptos_framework);
+        table::contains(&voting_records.votes, record_key)
+    }
+
+    #[view]
+    /// Return remaining voting power of a stake pool on a proposal.
+    /// Note: a stake pool's voting power on a proposal could increase overtime(e.g. rewards/new stake).
+    public fun get_remaining_voting_power(stake_pool: address, proposal_id: u64): u64 acquires VotingRecords, VotingRecordsV2 {
+        assert_voting_initialization();
+
+        // The voter's stake needs to be locked up at least as long as the proposal's expiration.
+        let proposal_expiration = voting::get_proposal_expiration_secs<GovernanceProposal>(@aptos_framework, proposal_id);
+        let lockup_until = stake::get_lockup_secs(stake_pool);
+        if (proposal_expiration > lockup_until) {
```

Comment:
When this can happen? Does this mean the stake_pool can only vote when its lockup_time > proposal_expiration?
n00b: how to make sure nobody can vote for an expired proposal?

Reply:
Yes a stake pool can only vote([ref](https://github.com/aptos-labs/aptos-core/blob/e5a981cdf48c8bcac65a504b2956230e913517c7/aptos-move/framework/aptos-framework/sources/aptos_governance.move#L322)).